core/src/main/java/com/yahoo/oak/NovaMemoryManager.java

```diff
@@ -0,0 +1,550 @@
+        }
+
+        /**
+         * Releases a write lock
+         *
+         * @return {@code TRUE} if the write lock was released successfully
+         * {@code FALSE} if the value is marked as deleted
+         * {@code RETRY} if the value was moved, or the version of the off-heap value does not match {@code version}.
+         */
+        public ValueUtils.ValueResult unlockWrite() {
+            assert version != ReferenceCodecSyncRecycle.INVALID_VERSION;
+            return HEADER.unlockWrite(version, getMetadataAddress(),  tap);
+        }
+
+        /**
+         * Marks the associated off-heap cut as deleted only if the version of that value matches {@code version}.
+         *
+         * @return {@code TRUE} if the value was marked successfully
+         * {@code FALSE} if the value is already marked as deleted
+         * {@code RETRY} if the value was moved, or the version of the off-heap value does not match {@code version}.
+         */
+        public ValueUtils.ValueResult logicalDelete() {
+            assert version != ReferenceCodecSyncRecycle.INVALID_VERSION;
+            return HEADER.logicalDelete(version, getMetadataAddress());
+        }
+
+        /**
+         * Is the associated off-heap cut marked as logically deleted
+         *
+         * @return {@code TRUE} if the value is marked
+         * {@code FALSE} if the value is not marked
+         * {@code RETRY} if the value was moved, or the version of the off-heap value does not match {@code version}.
+         */
+        public ValueUtils.ValueResult isDeleted() {
+            assert version != ReferenceCodecSyncRecycle.INVALID_VERSION;
+            return HEADER.isLogicallyDeleted(version, getMetadataAddress());
+        }
+
+        /**
+         * Marks the header of the associated off-heap cut as moved, just write (without CAS)
+         * The write lock must be held (asserted inside the header)
+         */
+        public void markAsMoved() {
+            assert associated;
+            HEADER.markAsMoved(getMetadataAddress());
+            throw new IllegalAccessError();
+        }
+
+        /**
+         * Marks the header of the associated off-heap cut as deleted, just write (without CAS)
+         * The write lock must be held (asserted inside the header).
+         * It is similar to logicalDelete() but used when locking and marking don't happen in one CAS
+         */
+        public void markAsDeleted() {
+            assert version != ReferenceCodecSyncRecycle.INVALID_VERSION;
+            HEADER.logicalDelete(version, getMetadataAddress());
+        }
+
+    }
+    @Override
+    public <K> int compareKeyAndSerializedKey(K key, OakScopedReadBuffer serializedKey, OakComparator<K> cmp) {
+        if (((ScopedReadBuffer) serializedKey).s.lockRead() != ValueResult.TRUE) {
+            throw new DeletedMemoryAccessException();
+        }
+        int res = cmp.compareKeyAndSerializedKey(key, serializedKey);
```

Comment:
This should be wrapped in try/catch to avoid locking the object forever.
The same issue is below and in other MMs.

Reply:
Hmm, again Nova is non blocking we don't really use locks,
in other MM however how could we lock object forever?

Reply:
If the object is locked, then the compare function fails with an exception, then we never call unlock.

Reply:
If we started the comparison, this means that we already successfully locked the key, we can not fail while comparing.
If we failed because we could not lock the key, then we throw the exception.

Reply:
`cmp.compareKeyAndSerializedKey(key, serializedKey);` is a user-provided function. You don't know what it does. Specifically, it might throw an unchecked exception. Then, unlock will not be called.